File: src/KbUtil.Lib/SvgGeneration/Internal/TextWriter.cs
namespace KbUtil.Lib.SvgGeneration.Internal
{
    using KbUtil.Lib.Models.Keyboard;
    using KbUtil.Lib.Extensions;
    using System.Collections.Generic;
    using System.Xml;

    internal class TextWriter : IElementWriter<Text>
    {
        private static readonly string DefaultTextAnchor = "middle";
        private static readonly string DefaultFont = "30px sans-serif";
        private static readonly string DefaultFill = "#000000";

        public SvgGenerationOptions GenerationOptions { get; set; }

        public void Write(XmlWriter writer, Text text)
        {
            writer.WriteStartElement("g");

            var elementWriter = new ElementWriter { GenerationOptions = GenerationOptions };

            elementWriter.WriteAttributes(writer, text);
            WriteAttributes(writer, text);

            elementWriter.WriteSubElements(writer, text);
            WriteSubElements(writer, text);

            writer.WriteEndElement();
        }

        public void WriteAttributes(XmlWriter writer, Text text)
        {
        }

        public void WriteSubElements(XmlWriter writer, Text text)
        {
            var styleDictionary = new Dictionary<string, string>();
            
            styleDictionary["font"] = !string.IsNullOrEmpty(text.Font)
                ? text.Font
                : DefaultFont;
            styleDictionary["fill"] = !string.IsNullOrEmpty(text.Fill)
                ? text.Fill
                : DefaultFill;

            string textAnchor = !string.IsNullOrEmpty(text.TextAnchor)
                ? text.TextAnchor
                : DefaultTextAnchor;
            
            if (GenerationOptions != null && GenerationOptions.EnableVisualSwitchCutouts == true)
            {
                writer.WriteStartElement("text");
                //writer.WriteAttributeString("id", "TODO");
                writer.WriteAttributeString("text-anchor", textAnchor);
                writer.WriteAttributeString("style", styleDictionary.ToCssStyleString());
                writer.WriteString(text.Content);
                writer.WriteEndElement();
            }
        }
    }
}

File: kb/pkg/svg/text_writer.go
package svg

import (
	"kb/pkg/models"
)

const (
	DefaultTextAnchor = "middle"
	DefaultFont       = "30px sans-serif"
	DefaultFill       = "#000000"
)

type textWriter struct {
	options *Options
}

func newTextWriter(options *Options) *textWriter {
	return &textWriter{options: options}
}

func (tw *textWriter) write(w *xmlWriter, text *models.Text) {
	if !text.Visible {
		return
	}

	w.writeStartElement("g")

	_elementWriter := newElementWriter(tw.options)

	_elementWriter.writeAttributes(w, text)
	tw.writeAttributes(w, text)

	_elementWriter.writeSubElements(w, text)
	tw.writeSubElements(w, text)

	w.writeEndElement()
}

func (tw *textWriter) writeAttributes(w *xmlWriter, text *models.Text) {
}

func (tw *textWriter) writeSubElements(w *xmlWriter, text *models.Text) {
	if !tw.options.EnableVisualSwitchCutouts {
		return
	}

	w.writeStartElement("text")
	w.writeAttributeString("text-anchor", stringOrDefault(text.TextAnchor, ""))
	w.writeAttributeString("style", cssStyleString(map[string]string{
		"font": stringOrDefault(text.Font, DefaultFont),
		"fill": stringOrDefault(text.Fill, DefaultFill),
	}))
	w.writeText(text.Content)
	w.writeEndElement()
}
